Add JsonFileSessionStore

Laiteux opened the pull request:
![JSON session file preview](https://share.laiteux.dev/jjvfkkct)

Laiteux commented:
Let me know what you think.

Laiteux commented:
Note that you have to add `Newtonsoft.Json` as a reference to the startup project (the one you're using TgSharp with), however this error occurs:

```
Exception levée : 'System.IO.FileNotFoundException' dans Newtonsoft.Json.dll
Exception levée : 'System.IO.FileNotFoundException' dans System.Private.CoreLib.dll
Exception levée : 'System.IO.FileNotFoundException' dans System.Private.CoreLib.dll
Une exception de type 'System.IO.FileNotFoundException' s'est produite dans System.Private.CoreLib.dll mais n'a pas été gérée dans le code utilisateur
Could not load file or assembly 'System.Security.Permissions, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. Le fichier spécifié est introuvable.
```

This might be annoying. After a few Google searches, I tend to believe the issue is due to the fact we recently migrated to .NET Standard 2.0: https://stackoverflow.com/a/48043024/12753809

knocte commented:
> we recently migrated to .NET Standard 2.0 or .NET Core

who is 'we'? I don't understand this sentence, can you clarify

Laiteux commented:
> > we recently migrated to .NET Standard 2.0 or .NET Core
> 
> who is 'we'? I don't understand this sentence, can you clarify

"we" = TgSharp. Nevermind tho, I used the master branch and it wasn't migrated yet. Forget about that.

knocte commented:
>  it wasn't migrated to .NET Standard 2.0 yet. Forget about that.

So you mean the System.IO.FileNotFoundException problem will only happen when we change TgSharp to .NETStandard?

Laiteux commented:
> So you mean the System.IO.FileNotFoundException problem will only happen when we change TgSharp to .NETStandard?

No, it **currently is** happening. I thought that was because we migrated to .NET Standard but I was wrong. Maybe migrating it will fix the issue?

knocte commented:
I see, ok.

BTW you haven't said why is this JsonFileSessionStore useful at all? Why did you include it? What benefit does it bring to you?

Laiteux commented:
> BTW you haven't said why is this JsonFileSessionStore useful at all? Why did you include it? What benefit does it bring to you?

I mean, it's always better to have a human-readable JSON file than an unreadable DAT one containing bytes.

You can see what the session file contains, what it is all about.

Also, it prevents the file from getting corrupted. Happened to me a few times in the past using `FileSessionStore`, had to delete the session file and reconnect/auth.

And it's just way better in general for storing data, in my opinion. For example, I currently store a lot of sessions in a SQL database, and doing it with bytes would have been a pain.

knocte commented:
I agree, thanks for the explanation. I'll be reviewing this PR today.

knocte commented:
2 questions:

1. Why is the commit `Set C# language version to 8.0 for all projects` needed?
2. Along with the changes of commit `Use JsonFileSessionStore as default TelegramClient ISessionStore`, can you provide a migration path so that, if a .dat file is found, it is converted to json first, and then used?

Laiteux commented:
1. I wanted to use the `??=` operator, which was only available from C# 8.0. I then thought upgrading the LangVersion for all projects wouldn't be a bad thing.
2. Can you please provide more explanations? What exactly do you want?

knocte commented:
> I wanted to use the ??= operator

Ah, do you actually mind to *not* use this operator please? I find it quite unreadable.


> Can you please provide more explanations? What exactly do you want?

Imagine some software developer uses TgSharp current master branch, and in the future he upgrades to a version that has this feature merged. If we merge this PR as is, what I guess will happen is that his program won't try to open the .dat file, instead it will try to open a saved JSON store file, it will not find it, and create a new one. However, what should happen is that it should be backwards compatible: TgSharp should see the existence of the .dat file (and no existence of JSON store file), then it should convert the .dat file to JSON, and afterwards, load the session from that JSON file.

Laiteux commented:
> Ah, do you actually mind to not use this operator please? I find it quite unreadable.

Ok, sure. Should I then revert 306dc8f8b76c8490e5f92eaf6eb288760d073bf4 too?

> Imagine some software developer uses TgSharp current master branch, and in the future he upgrades to a version that has this feature merged...

Got it, will write that tomorrow.

knocte commented:
> Should I then revert 306dc8f too?

Yup because it's not needed anymore. We will upgrade C# when we really need it.

Laiteux commented:
Tested & working 👌

knocte commented:
Cool thanks! BTW, are you good at rebasing, separating commits et al?

Laiteux commented:
Never done that. Honestly this is one of my first "serious" pull requests, I'm still a bit of a beginner with git.

knocte commented:
So if I ask you to separate your last commit in 2 commits (one that removes the ??= operator, and one that introduces the migration), will you know how to do it?

Laiteux commented:
I have no idea if I did it the way you wanted, but here we go

knocte commented:
There's a merge commit, so no :( you should have used interactive rebase

Laiteux commented:
I have no idea of how to do that, sorry

knocte commented:
If you give me access to your fork I'll remove the merge commit

Laiteux commented:
https://github.com/Laiteux/TgSharp/invitations

knocte commented:
@Laiteux see how I reordered commits, squashed some of them and adjusted commit messages (thanks to the magic of `git rebase -i`)

Laiteux commented:
Looks cool!

knocte commented:
Thanks. By the way I've just realised there may be a couple more improvements to make, I'm going to point them out now, but be careful, given that I pushed with `--force` into your master branch, I'd recommend to work on a new git clone next.

knocte commented:
(Otherwise, you may get conflicts, or git merge commits, if you pull.)

knocte commented on src/TgSharp.Core/DataCenter.cs:
```diff
@@ -1,20 +1,25 @@
+    public class DataCenter
     {
-        internal DataCenter (int? dcId, string address, int port)
+        public DataCenter()
```
given that this constructor will only be used when serializing/deserializing, I think we should add an `[Obsolete("do not use, this ctor is public only for serialization"]` attribute

knocte commented on src/TgSharp.Core/MTProto/Crypto/AuthKey.cs:
```diff
@@ -9,6 +9,12 @@ public class AuthKey
         private ulong keyId;
         private ulong auxHash;
+
+        public AuthKey()
```
same here?

knocte commented on src/TgSharp.Core/packages.config:
```diff
@@ -1,3 +1,4 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Newtonsoft.Json" version="12.0.3" targetFramework="net462" />
```
please note: you're adding v12 here but v9 in other project

knocte commented on src/TgSharp.Core/Session.cs:
```diff
@@ -87,14 +138,19 @@ public class Session
+        [JsonIgnore]
+        public ISessionStore Store;
 
-        private ISessionStore store;
```
is this store->Store change really needed? why make it public if you're adding the `[JsonIgnore]` attrib to it anyway?

Laiteux replied:
I made it public so if a user wants to make his own `ISessionStore`, he will be able to set the `Store` property if needed, like [here](https://github.com/Laiteux/TgSharp/blob/master/src/TgSharp.Core/Session.cs#L101) for example.

knocte replied:
but there's already a Session ctor that receives an ISessionStore instance, you could have used that instead

Laiteux replied:
Remember, you can only deserialize JSON to parameterless constructors

knocte replied:
I know? you did this:

```
            var session = JsonConvert.DeserializeObject<Session>(File.ReadAllText(sessionPath), jsonSerializerSettings);

            session.Store = this;
            session.SessionUserId = sessionUserId;

            return session;
```

why not this?:

```
            var deserializedSession = JsonConvert.DeserializeObject<Session>(File.ReadAllText(sessionPath), jsonSerializerSettings);
            var newSession = Session(deserializedSession);
            newSession.SessionUserId = sessionUserId;

            return newSession;
```

Laiteux replied:
That requires creating 2 instances of the `Session` class, I prefer my current code, also easier to read imo

Laiteux replied:
Wait, actually your code is invalid, an `ISessionStore` must be passed to the `Session` ctor, what are you trying to achieve?

knocte replied:
mmm sorry I'm confused now; i just don't understand why the binary store didn't need this field to be public, while yours needs it; public mutable fields are evil, I'm sure it can be done differently

Laiteux replied:
Looks like it is because `FileSessionStore` ("binary store") is setting it using the ctor. This might be able to be done differently (not so sure), but trust me we still want this property to be public. I use it for my SQL store I talked to you about a few hours ago, for example, it is required. And I don't think it's a bad thing at all to publicly expose it anyway.

knocte replied:
>  is setting it using the ctor

Let's do the same then.


> I use it for my SQL store

Maybe for your SQL store it's enough to make it protected?

Laiteux replied:
> Let's do the same then.

That would require creating a second instance of the `Session` class, only for using the ctor. I don't think this is a good practice. What's so wrong with the current code?

> Maybe for your SQL store it's enough to make it protected?

Should be. You decide.

knocte replied:
> Should be. You decide.

Let's use `protected` then, public fields are the classic way to shoot yourself in the foot. Creating second instances is not a big deal, our machines are very beefy these days (loss of performance better than loss of maintainability, developer time is more expensive than CPU cycles).

Laiteux replied:
I just realized that by creating a new `Session` instance, all property values we got via deserialization will be lost.

What I mean is that we would then have to do that:

```cs
var deserializedSession = JsonConvert.DeserializeObject<Session>(File.ReadAllText(sessionPath), jsonSerializerSettings);

var session = new Session(this)
{
    AuthKey = deserializedSession.AuthKey,
    DataCenter = deserializedSession.DataCenter,
    ...
};

return session;
```

Not convenient.

knocte replied:
> we got via deserialization will be lost.

why this doesn't happen in the binary store? should we do a new ctor that lets you do `Session(deserializedSession, this)`?

Laiteux replied:
Because the binary store uses this:
```cs
return new Session(store)
{
    AuthKey = new AuthKey(authData),
    Id = id,
    Salt = salt,
    Sequence = sequence,
    LastMessageId = lastMessageId,
    TimeOffset = timeOffset,
    SessionExpires = sessionExpires,
    TLUser = TLUser,
    SessionUserId = sessionUserId,
    DataCenter = defaultDataCenter,
};
```
🤮

knocte replied:
haha

knocte replied:
actually, that code from the binary store would benefit from the new ctor that I proposed: `Session(deserializedSession, this)`, this way, the vomitive property mapping is only centralized in one place instead of on each store type

Laiteux replied:
I thought the same, except the binary vomitive property mapping doesn't use a `Session` instance but this:
```cs
var id = reader.ReadUInt64();
var sequence = reader.ReadInt32();
var salt = reader.ReadUInt64();
var lastMessageId = reader.ReadInt64();
var timeOffset = reader.ReadInt32();
var serverAddress = Serializers.String.Read(reader);
var port = reader.ReadInt32();
```
😒

Laiteux replied:
Back to the beginning. 😕

knocte replied:
no! you showed me the code before:

```
    AuthKey = new AuthKey(authData),
    Id = id,
    Salt = salt,
    Sequence = sequence,
    LastMessageId = lastMessageId,
    TimeOffset = timeOffset,
    SessionExpires = sessionExpires,
    TLUser = TLUser,
    SessionUserId = sessionUserId,
    DataCenter = defaultDataCenter,
```

that code above would then be in the new ctor

Laiteux replied:
Yes! But then how do you call the new ctor from `Session.FromBytes`?

knocte replied:
(used by both the binary and the json store)

knocte replied:
holy shit, after doing so much F# programming, I'm starting to realise how much I hate OOP

Laiteux replied:
LMAO

knocte commented on src/TgSharp.Core/TelegramClient.cs:
```diff
@@ -57,14 +58,16 @@ public Session Session
-            this.store = store;
+            this.store = store ?? new JsonFileSessionStore();
+
+            if (store == null && File.Exists($"{sessionUserId}.dat"))
```
I think this is a bit confusing because of the name clash between the parameter `store` and the field `this.store`, it would be more clear if we do this:

```
if (store != null) {
    this.store = store;
} else {
    this.store = new JsonFileSessionStore();
    if (File.Exists($"{sessionUserId}.dat")) {
        MigrateOldSessionFile(sessionUserId);
    }
}
```

Laiteux replied:
Agreed, it is confusing, would've been better if this was using `_camelCase` naming for private fields 😝

Laiteux replied:
Switched to your code tho, except I inverted the `if` statement. Way cleaner and less confusing. Let me know if there's anything else you'd like to discuss about, then I'll push.

knocte replied:
ah right, if you do it inverted, you can even save one line:

```
if (store == null) {
    store = new JsonFileSessionStore();
    if (File.Exists($"{sessionUserId}.dat")) {
        store = MigrateOldSessionFile(sessionUserId);
    }
}
this.store = store;
```

Laiteux replied:
You can't, `MigrateOldSessionFile` uses `this.store` so it must be set before we call it. Nice try tho 😄

Laiteux replied:
Unless you want to do that inside of `MigrateOldSessionFile`, and return an `ISessionStore` from it?

Laiteux replied:
```cs
if (store == null && File.Exists($"{sessionUserId}.dat"))
{
    store = MigrateOldSessionFile(sessionUserId);
}

this.store = store;
```
```cs
private ISessionStore MigrateOldSessionFile(string sessionUserId)
{
    var jsonSessionStore = new JsonFileSessionStore();
    var fileSessionStore = new FileSessionStore();

    var session = fileSessionStore.Load(sessionUserId);

    jsonSessionStore.Save(session);

    File.Delete($"{sessionUserId}.dat");

    return jsonSessionStore;
}

```

Laiteux replied:
Like that:
```cs
if (store == null)
{
    store = File.Exists($"{sessionUserId}.dat") ? MigrateOldSessionFile(sessionUserId) : new JsonFileSessionStore();
}

this.store = store;
```
```cs
private ISessionStore MigrateOldSessionFile(string sessionUserId)
{
    var jsonSessionStore = new JsonFileSessionStore();
    var fileSessionStore = new FileSessionStore();

    var session = fileSessionStore.Load(sessionUserId);

    jsonSessionStore.Save(session);

    File.Delete($"{sessionUserId}.dat");

    return jsonSessionStore;
}
```
Doesn't seem better tho.

knocte replied:
I like it more, and you can make MigrateOldSessionFile static

Laiteux replied:
Let's use this:
```cs
if (store == null)
{
    this.store = new JsonFileSessionStore();

    if (File.Exists($"{sessionUserId}.dat"))
    {
        MigrateOldSessionFile(sessionUserId);
    }
}
else
{
    this.store = store;
}
```
```cs
private void MigrateOldSessionFile(string sessionUserId)
{
    var fileSessionStore = new FileSessionStore();

    var session = fileSessionStore.Load(sessionUserId);

    store.Save(session);

    File.Delete($"{sessionUserId}.dat");
}
```

knocte replied:
this also makes me think that FilleSessionStore.Load should be static

Laiteux replied:
It can't be static? `Load` is a method from the `ISessionStore` interface

Laiteux commented:
It's midnight, I'm getting super tired.

I pushed my current working version of the code.

We debated for more than an hour on this session stuff, so if you come up with an idea feel free to commit it yourself.

Going to sleep, it was a pleasure talking with you, really enjoyed it!

Talk to you tomorrow 😊

knocte commented:
It's only 10PM over here 😅 , I might give this some extra thought.

knocte commented:
After some thought, I've realised that possibly the wrong thing that the current code was doing is storing the store as a field of the Session class, so I've removed it, refactoring the code a little bit, in this experiment: e6531281721363cbb73f7b438d687c6965bb00cb . Can you include this commit at the beginning of this PR when you have some time?

Laiteux commented:
Not so sure about how to do that.

knocte commented:
I can teach you but I think it's gonna be simpler if you first squash your latest 4 commits with the 2 initial ones. Use `git rebase -i HEAD~6` to do that first, then push with --force.

Laiteux commented:
```
C:\Users\matt1\source\repos\TgSharp>git rebase -i HEAD~6
Successfully rebased and updated refs/heads/master.

C:\Users\matt1\source\repos\TgSharp>git push --force
Everything up-to-date
```
Looks like nothing happened

knocte commented:
wtf, `git rebase -i HEAD~6` should open an editor; when you do `git commit` don't you get an editor opened?

Laiteux commented:
> wtf, `git rebase -i HEAD~6` should open an editor; when you do `git commit` don't you get an editor opened?

Yes, then I close it and `Successfully rebased and updated refs/heads/master.` shows up.

However, after using `git push --force`, nothing happens: `Everything up-to-date`

knocte commented:
> then I close it 

Oh, well, you shouldn't just close it. That editor allows you to: reorder commits (if you move lines up or down), squash commits (if you replace a word 'pick' with 's'), or reword commits (if you replace a word 'pick' with 'r').

Laiteux commented:
Oh, ok. I didn't know about that, I've always been using GitHub desktop without any advanced "commands", I guess I should take a course on git.

So how would I do that in the editor? I honestly have no idea, never have done that before.
> squash your latest 4 commits with the 2 initial ones

knocte commented:
Ok you should see this:

```
pick c4e2c63 Add a JSON-format file session store
pick 5edf67d Use JsonFileSessionStore as default TelegramClient ISessionStore
pick 3352714 Fixed Newtonsoft.Json path & updated to latest version
pick 634f331 Added ObsoleteAttributes to DataCenter & AuthKey ctors
pick a1a592e Made the old session file migration process cleaner and less confusing
pick f52c0c1 Added ObsoleteAttribute to Session parameterless ctor
```

First let's think how we want to do it: I guess we want the 3rd, 4th and 6th commits merged with the 1st and the 5th commit merged with the second, then move lines up to end up like this:

```
pick c4e2c63 Add a JSON-format file session store
pick 3352714 Fixed Newtonsoft.Json path & updated to latest version
pick 634f331 Added ObsoleteAttributes to DataCenter & AuthKey ctors
pick f52c0c1 Added ObsoleteAttribute to Session parameterless ctor
pick 5edf67d Use JsonFileSessionStore as default TelegramClient ISessionStore
pick a1a592e Made the old session file migration process cleaner and less confusing
```

After that, you can mark 2nd, 3rd and 4th lines with an "f" (it's the same as 's', which stands for squash, but signaling that you just want the commit message to be the same as the 1st commit). So, like this:

```
pick c4e2c63 Add a JSON-format file session store
f 3352714 Fixed Newtonsoft.Json path & updated to latest version
f 634f331 Added ObsoleteAttributes to DataCenter & AuthKey ctors
f f52c0c1 Added ObsoleteAttribute to Session parameterless ctor
pick 5edf67d Use JsonFileSessionStore as default TelegramClient ISessionStore
f a1a592e Made the old session file migration process cleaner and less confusing
```

After doing that, you will have only 2 commits in your history, and you can push with force.

knocte commented:
I just saw that you did a merge, why did you merge? we're rebasing, not merging

Laiteux commented:
> I just saw that you did a merge, why did you merge? we're rebasing, not merging
That was a mistake, I reverted the commit, hopefully it's ok now.

Laiteux commented:
> After doing that, you will have only 2 commits in your history, and you can push with force.

This is really dope. Thanks for the lesson!

Laiteux commented:
> After doing that, you will have only 2 commits in your history, and you can push with force.

This is really dope. Thanks for the lesson!

knocte commented:
No prob! Ok, now we want to include my refactoring commit in this PR.

First, I guess you have two remotes when doing `git remote -v`, one should point to your fork and the other one to https://github.com/nblockchain/TgSharp. Let's say that the remote that points to your fork is called `laiteuxFork` and the latter is called `origin`. Then you do `git fetch origin` first, now you have my commit somewhere in your clone metadata. You know what my commit is in the branch `refactoring/Session` so you should just do `git rebase origin/refactoing/Session`: if the rebase can't be done cleanly, it will give some conflicts, but don't worry, just fix the conflicts in the files that contains strings such as `>>>>` or `<<<<` and after that you can `git add` each file, and do `git rebase --continue`, until the rebase is done.

Laiteux commented:
**Should** be good.

knocte commented:
> Should be good.

Yep, you successfully injected my commit!

knocte commented on src/TgSharp.Core/Session.cs:
```diff
@@ -107,11 +159,19 @@ public class Session
+        private readonly Random random = new Random();
+
+        [JsonIgnore]
+        public ISessionStore Store;
```
now, after my refactoing commit is there, I guess removing this field is much easier now

Laiteux commented:
Cool!

knocte commented on src/TgSharp.TL/TL/TLUser.cs:
```diff
@@ -40,8 +40,8 @@ public override int Constructor
-        public TLAbsUserProfilePhoto Photo { get; set; }
-        public TLAbsUserStatus Status { get; set; }
+        [JsonIgnore] public TLAbsUserProfilePhoto Photo { get; set; }
+        [JsonIgnore] public TLAbsUserStatus Status { get; set; }
```
I've realised this is an autogenerated file, so whenever we upgrade the TL Layer, we would lose these changes by accident; I'm wondering, can we create a new derived class that inherits from TLUser and marks these two properties with [JsonIgnore]?

Laiteux replied:
What do you mean? I'm not sure about understanding your idea. Can you go more in depth?

knocte replied:
```
public class UserForSession : TLUser {
    [JsonIgnore]
    public new TLAbsUserProfilePhoto Photo { get; set; }
    [JsonIgnore]
    public new TLAbsUserStatus Status { get; set; }
}
```

Laiteux replied:
Yes, that would probably work. Will try and commit this after your fix.

Laiteux replied:
Update: See latest commit message

Laiteux commented:
Everything is perfectly working as intended. Good job! 🎉🎉🎉

AndersMad commented:
The TLUser got lost? Isn't that the only way for knowing the current connected/authed user id?

Laiteux commented:
> The TLUser got lost? Isn't that the only way for knowing the current connected/authed user id?

That is right. Even tho `TelegramClient.MakeAuthAsync` returns a `TLUser` object, if we use an already authed session, you're right on the point I wouldn't really know how to get the user id or info in general.

@knocte

knocte commented:
Oh, we pruned too much. TBH I didn't realise that Session might be of use to the TgSharp consumer too (it only recently became a public property thanks to a pull request contributed by someone). Maybe we should just store userID? @AndersMad what are the TLUser properties that you use? we don't want to store the whole TLUser in the session

Laiteux commented:
I believe storing the user ID should be enough, and if the consumer really wants all the other info linked to the TLUser, he's free to store them in his own database, using the user ID as key.

Laiteux commented:
But yeah, I also was using the Session property in my own program, and had to refactor it a little bit since we removed it. I actually believe it might be useful, I'm not sure if there are any case in which it would be indispensable tho.

knocte commented:
Ok sounds good, do you mind changing it Matt.

Laiteux commented:
No problem, but that still won't expose TLUser.

knocte commented:
> why didn't we simply use the JsonIgnoreAttribute

when you don't remember something, you can still find it in the github UI ;) : https://github.com/nblockchain/TgSharp/pull/18#discussion_r488403038

Laiteux commented:
So why didn't we simply do that?:
```cs
[JsonIgnore]
public TLUser TLUser { get; set; }
```

Laiteux commented:
However, that would mean it STILL would be included in the binary session file, for example. Let me know what you think

knocte commented:
We don't want the binary session file to store different things than the JSON session file.

Laiteux commented:
So what should we do? 😕

Laiteux commented:
User could still get TLUser using MakeAuthAsync and store it himself.

knocte commented:
> User could still get TLUser using MakeAuthAsync and store it himself.

So, you meant to answer "no" to Anders' below question?:

> Isn't that the only way for knowing the current connected/authed user id?

Laiteux commented:
Yes, this is what I meant.

AndersMad commented:
Ye, I just added the id from TLUser to the Session persisting (1 line in session and 1 line in client on auth) - the other (TLUser) values are flux anyways - but the user id is nice to know.

Laiteux commented:
Sorry, what do you mean by that? I don't understand

> Ye, I just added the id from TLUser to the Session persisting (1 line in session and 1 line in client on auth)

Laiteux commented:
Nevermind, got it. Would you mind contributing to this PR?

AndersMad commented:
@Laiteux In my code to make it work.. I just added `UserId` prop to the session so it gets persisted and then set it where the TLUser was set in the Cient.

AndersMad commented:
@Laiteux Yes, but I'm so so bad at git/github - I always f'up something :P .. But was just this + added the prop to `Session`
```
private void OnUserAuthenticated(TLUser TLUser) {
	Session.UserId = TLUser.Id;
```

Laiteux commented:
That's fine, I'll add it then rebase. Ok for you @knocte?

knocte commented:
yes sounds fine; note you may have problems injecting that change in the 2nd commit because the 3rd commit changes stuff around that too (if this is the case, you can just merge the 2nd and 3rd commits into 1, copy+pasting the text from the commit message from the 3rd)

aarani commented:
No problem if you guys want to remove TLUser but you have to have a GetMe function to get the TLUser object ( if user == null (can be overruled with a force_update arg) )

knocte commented:
> ( if user == null (can be overruled with a force_update arg) )

what do you mean with the above @aarani ?

aarani commented:
`class Session
{
   internal TLUser TLUser;
}

class TelegramClient
{
   void MakeAuthAsync()
   {
      session.TLUser = authResult.TLUser; // pseudo-code
   }

   async Task<TLUser> GetMe(bool refresh = false)
   {
      if (refresh || session.TLUser == null)
      {
         session.TLUser = await this.GetUser(self); // pseudo-code
      }

      return session.TLUser;
   }
}`

knocte commented:
I rebased this PR in this branch: https://github.com/nblockchain/TgSharp/commits/wip/jsonSessionStore , can you review @aarani ? (Note: CI might be broken in the last commit, I just have to replace the .dat file with a new .json file, I'll do that tomorrow.)

aarani commented:
> I rebased this PR in this branch: https://github.com/nblockchain/TgSharp/commits/wip/jsonSessionStore , can you review @aarani ? (Note: CI might be broken in the last commit, I just have to replace the .dat file with a new .json file, I'll do that tomorrow.)

LGTM but it probably needs more testing

Laiteux committed:
Use JSON as default file session store

It's always better to have a human-readable JSON file than
an unreadable DAT one containing bytes.

You can see what the session file contains, what it is all
about.

Also, it prevents the file from getting corrupted. Happened
to me a few times in the past using FileSessionStore, had
to delete the session file and reconnect/auth.

And it's just way better in general for storing data. For
example, I currently store a lot of sessions in a SQL
database, and doing it with bytes would have been a pain.

JSON session file is just better in general.

NOTE: Reason for migrating to .NET4.6.2 is to be able to
reference .NETStandard2.0 (needed for System.Text.Json).

Co-authored-by: Andres G. Aragoneses <knocte@gmail.com>

knocte commented:
Migrated to System.Text.Json and fixed binary serialization nits pointed by @aarani.
Hey @Laiteux by any chance can you test this before I merge?

omerjacobi commented:
what is still missing to finish this pull request?

Laiteux commented:
Nothing I believe, iirc it works well but we wanted to do some more tests before merging, which we never ended up doing

omerjacobi commented:
> Nothing I believe, iirc it works well but we wanted to do some more tests before merging, which we never ended up doing

Is there anything I can help with? I really like to use this feature in the official NuGet

Laiteux commented:
Please test the feature on your side and let us know if everything works fine. If so, @knocte will most likely merge it

knocte commented:
> Is there anything I can help with? 

@omerjacobi you wanna contribute? tell me your username in telegram and I'll add you to a group where we can chat

omerjacobi commented:
> > Is there anything I can help with?
> 
> @omerjacobi you wanna contribute? tell me your username in telegram and I'll add you to a group where we can chat
 
@knocte I send it to your gmail

Laiteux committed:
Trigger build
